expr/func_aspercent.go

```diff
@@ -0,0 +1,268 @@
+					serie1.Datapoints[i].Val = computeAsPercent(serie1.Datapoints[i].Val, serie2.Datapoints[i].Val)
+				}
+				outSeries = append(outSeries, serie1)
+			}
+			return outSeries, nil
+		} else {
+			return nil, errors.New("asPercent second argument must be missing, a single digit, reference exactly 1 series or reference the same number of series as the first argument")
+		}
+	} else {
+		totalsSerie.QueryPatt = fmt.Sprint(s.totalFloat)
+	}
+
+	for _, serie := range series {
+		serie.QueryPatt = fmt.Sprintf("asPercent(%s,%s)", serie.QueryPatt, totalsSerie.QueryPatt)
+		serie.Target = fmt.Sprintf("asPercent(%s,%s)", serie.Target, totalsSerie.QueryPatt)
+		serie.Tags = map[string]string{"name": serie.Target}
+		deepCopySerieElements(&serie)
+		for i := range serie.Datapoints {
+			var totalVal float64
+			if len(totalsSerie.Datapoints) > i {
+				totalVal = totalsSerie.Datapoints[i].Val
+			} else {
+				totalVal = s.totalFloat
+			}
+			serie.Datapoints[i].Val = computeAsPercent(serie.Datapoints[i].Val, totalVal)
+		}
+		outSeries = append(outSeries, serie)
+	}
+	return outSeries, nil
+}
+
+func computeAsPercent(in, total float64) float64 {
+	if math.IsNaN(in) || math.IsNaN(total) {
+		return math.NaN()
+	}
+	if total == 0 {
+		return math.NaN()
+	}
+	return in / total * 100
+}
+
+func groupSeriesByKey(series []models.Series, nodes []expr, keys *map[string]struct{}) map[string][]models.Series {
+	keyedSeries := make(map[string][]models.Series)
+	for _, serie := range series {
+		key := aggKey(serie, nodes)
+		if _, ok := keyedSeries[key]; !ok {
+			keyedSeries[key] = []models.Series{serie}
+			(*keys)[key] = struct{}{}
+		} else {
+			keyedSeries[key] = append(keyedSeries[key], serie)
+		}
+	}
+	return keyedSeries
+}
+
+// Sums each seriesList in map of seriesLists
+func getTotalSeries(totalSeriesLists map[string][]models.Series) map[string]models.Series {
+	totalSeries := make(map[string]models.Series, len(totalSeriesLists))
+	for key := range totalSeriesLists {
+		totalSeries[key] = sumSeries(totalSeriesLists[key])
+	}
+	return totalSeries
+}
+
+// Sums seriesList
+// Datapoints are always a copy
+func sumSeries(series []models.Series) models.Series {
+	if len(series) == 1 {
+		deepCopySerieElements(&series[0])
+		return series[0]
+	}
+	out := pointSlicePool.Get().([]schema.Point)
+	crossSeriesSum(series, &out)
+	var queryPatts []string
+
+Loop:
+	for _, v := range series {
+		// avoid duplicates
+		for _, qp := range queryPatts {
+			if qp == v.QueryPatt {
+				continue Loop
+			}
+		}
+		queryPatts = append(queryPatts, v.QueryPatt)
+	}
+	name := fmt.Sprintf("sumSeries(%s)", strings.Join(queryPatts, ","))
+	cons, queryCons := summarizeCons(series)
+	return models.Series{
+		Target:       name,
+		QueryPatt:    name,
+		Datapoints:   out,
+		Interval:     series[0].Interval,
+		Consolidator: cons,
+		QueryCons:    queryCons,
+		Tags:         map[string]string{"name": name},
+	}
+}
+
+func deepCopySerieElements(serie *models.Series) {
```

Comment:
this API is a bit strange. would it not make more sense to return a new series that is a deep copy?
even though that is slightly more work, seems worth it.

in fact, maybe add a `Copy()` method to the serie type in the models package

Reply:
also looking at sumSeries, the copying/newly-allocating seems a bit too eager.
we should only need to allocate point slices:
1) for the totals slice: when we need to sum up values and we need a place to store the totals (not when totals results in a single series, then we can just read from it)
2) for each output series.(if it is different from the input slice)

Reply:
The problem with #1 is that later I modify the total series in some cases such as this (line 99):

```
// No series for total series
if _, ok := metaSeries[key]; !ok {
	serie2 := totalSeries[key]
	serie2.QueryPatt = fmt.Sprintf("asPercent(MISSING,%s)", serie2.QueryPatt)
	serie2.Target = fmt.Sprintf("asPercent(MISSING,%s)", serie2.Target)
	serie2.Tags = map[string]string{"name": serie2.Target}
	for i := range serie2.Datapoints {
		serie2.Datapoints[i].Val = math.NaN()
	}
	outSeries = append(outSeries, serie2)
	continue
}
```

Reply:
I'll be mostly afk until Thursday.
Please read expr/NOTES carefully if you have not already done so. All this complexity is to make sure we never overwrite data in the memory AggMetric, chunk cache etc.
Some of these fields are by value so harmless but the tags map is interesting as well. Not sure if we currently take that into account everywhere. On phone so can't check right now

Reply:
Ok, well I added a `Series.Copy(emptyDatapoints []schema.Point)` function (the argument is there so that `pointSlicePool` can be used if needed). And I only copy the series if I modify values that shouldn't be modified.

Hopefully that's sufficient and what you were looking for 😃